spanner/client.go

```diff
@@ -121,6 +124,131 @@ func (c *Client) ClientID() string {
+		cfg.GRPCgcpConfig.Method = []*grpcgcppb.MethodConfig{
+			{
+				Name: []string{"/google.spanner.v1.Spanner/CreateSession"},
+				Affinity: &grpcgcppb.AffinityConfig{
+					Command:     grpcgcppb.AffinityConfig_BIND,
+					AffinityKey: "name",
+				},
+			},
+			{
+				Name: []string{"/google.spanner.v1.Spanner/BatchCreateSessions"},
+				Affinity: &grpcgcppb.AffinityConfig{
+					Command:     grpcgcppb.AffinityConfig_BIND,
+					AffinityKey: "session.name",
+				},
+			},
+			{
+				Name: []string{"/google.spanner.v1.Spanner/DeleteSession"},
+				Affinity: &grpcgcppb.AffinityConfig{
+					Command:     grpcgcppb.AffinityConfig_UNBIND,
+					AffinityKey: "name",
+				},
+			},
+			{
+				Name: []string{"/google.spanner.v1.Spanner/GetSession"},
+				Affinity: &grpcgcppb.AffinityConfig{
+					Command:     grpcgcppb.AffinityConfig_BOUND,
+					AffinityKey: "name",
+				},
+			},
+			{
+				Name: []string{
+					"/google.spanner.v1.Spanner/BeginTransaction",
+					"/google.spanner.v1.Spanner/Commit",
+					"/google.spanner.v1.Spanner/ExecuteBatchDml",
+					"/google.spanner.v1.Spanner/ExecuteSql",
+					"/google.spanner.v1.Spanner/ExecuteStreamingSql",
+					"/google.spanner.v1.Spanner/PartitionQuery",
+					"/google.spanner.v1.Spanner/PartitionRead",
+					"/google.spanner.v1.Spanner/Read",
+					"/google.spanner.v1.Spanner/Rollback",
+					"/google.spanner.v1.Spanner/StreamingRead",
+				},
+				Affinity: &grpcgcppb.AffinityConfig{
+					Command:     grpcgcppb.AffinityConfig_BOUND,
+					AffinityKey: "session",
+				},
+			},
+		}
+	}
+	// Append emulator options if SPANNER_EMULATOR_HOST has been set.
+	if emulatorAddr := os.Getenv("SPANNER_EMULATOR_HOST"); emulatorAddr != "" {
+		emulatorOpts := []option.ClientOption{
+			option.WithEndpoint(emulatorAddr),
+			option.WithGRPCDialOption(grpc.WithTransportCredentials(insecure.NewCredentials())),
+			option.WithoutAuthentication(),
+			internaloption.SkipDialSettingsValidation(),
+		}
+		opts = append(opts, emulatorOpts...)
+		// Replace all endpoints with emulator target.
+		for _, meo := range cfg.MultiEndpoints {
+			meo.Endpoints = []string{emulatorAddr}
+		}
+	}
+
+	// Set the number of channels to the default value if not specified.
+	if cfg.GRPCgcpConfig.GetChannelPool() == nil || cfg.GRPCgcpConfig.GetChannelPool().GetMaxSize() == 0 {
+		cfg.GRPCgcpConfig.ChannelPool = &grpcgcppb.ChannelPoolConfig{
+			MinSize: numChannels,
+			MaxSize: numChannels,
+		}
+	}
+	// Set MinSize equal to MaxSize to create all the channels beforehand.
+	cfg.GRPCgcpConfig.ChannelPool.MinSize = cfg.GRPCgcpConfig.ChannelPool.GetMaxSize()
+
+	cfg.GRPCgcpConfig.ChannelPool.BindPickStrategy = grpcgcppb.ChannelPoolConfig_ROUND_ROBIN
+
+	cfg.DialFunc = func(ctx context.Context, target string, dopts ...grpc.DialOption) (*grpc.ClientConn, error) {
+		copts := opts
+
+		for _, do := range dopts {
+			copts = append(copts, option.WithGRPCDialOption(do))
+		}
```

Comment:
This could potentially be even worse.  If the `DialFunc` is run concurrently in the `ChannelPool`, then this function will be overwriting memory in one goroutine while reading it in another, which could cause serious problems, and should be expected to trigger the race detector if nothing else.  It looks like that doesn't happen today - grpc clients are created serially, but I could see a user complaining about startup performance and sending a pr that does them each in their own goroutine, and then this becomes broken as a result.  A copy should ideally be made before mutating the opts slice.

Reply:
Oops, that wasn't intended. Thank you for pointing this out. Sending PR #11151  to fix this.